# Fixed parameters for Mhsflt_toolkit pipeline

# Default configuration structure
DEFAULT_CONFIG = {
    "project": "",  # Will be set from command line
    "sample": {
        "sample_species": "human",  # Dummy value, user said species is not needed
        "sample_type": "auto",      # Will be set based on --manual
        "sample_id": "1"            # Will be set from --plate
    },
    "sequence_files": {
        "file1": {
            "name": "",             # Will be set from --fastq
            "base_definition": []   # Will be calculated based on read length
        },
        "file2": {
            "name": "",             # Will be set from --fastq
            "base_definition": []   # Will be calculated based on read length
        }
    },
    "reference": {
        "STAR_index": "",           # Will be set from --genomeDir
        "GTF_file": "",             # Will be detected in --genomeDir
        "additional_files": None,
        "additional_STAR_params": "--clip3pAdapterSeq CTGTCTCTTATACACATCT"
    },
    "out_dir": "",                  # Will be set to current_dir/sample_name
    "chemistry": "MGI",
    "num_threads": 30,              # Will be set from --threads
    "mem_limit": 0,                 # 0 means unlimited
    "filter_cutoffs": {
        "BC_filter": {
            "num_bases": 4,
            "phred": 5
        },
        "UMI_filter": {
            "num_bases": 3,
            "phred": 5
        }
    },
    "barcodes": {
        "barcode_num": None,
        "automatic": False,
        "BarcodeBinning": 1,
        "nReadsperCell": 1,
        "demultiplex": False,
        # barcode_file will be generated by the pipeline logic
        "barcode_file": "" 
    },
    "counting_opts": {
        "introns": True,
        "downsampling": "0",
        "strand": 1,
        "Ham_Dist": 1,
        "velocyto": False,
        "primaryHit": True,
        "twoPass": False
    },
    "make_stats": True,
    "make_sorted_bam": False,
    "make_ub_bam": False,
    "which_Stage": "Filtering",
    "zUMIs_directory": ".",         # Will be resolved relative to script
    "read_layout": "PE"             # Default PE, verified by script
}
